### ISSUE TITLE
Cannot create TTL index on multiple fields in MongoDB 4.0 (as defined on the README.md file)

### ISSUE THREAD

**Issue by cesarizu** [mongoid/mongoid-locker#91]
I have an index for a model that I want to use `mongoid-locker` with.

```
index({ _id: 1, locking_name: 1 }, name: 'mongoid_locker_index', sparse: true, unique: true, expire_after_seconds: lock_timeout)
```

When trying to create indexes I get this error:

```
Mongo::Error::OperationFailure:
  TTL indexes are single-field indexes, compound indexes do not support TTL. Index spec: { key: { _id: 1, locking_name: 1 }, name: "mongoid_locker_index", sparse: true, unique: true, expireAfterSeconds: 5 } (67) (on localhost:27017)
```

Should I be doing something different?

# Versions:

mongoid-locker=2.0.1
mongoid=6.4.2
mongodb=4.0

**Comment by dblock** [comment]
Hm, in the doc we recommend creating exactly that, don't we? And we have it in https://github.com/mongoid/mongoid-locker/blob/f00416d3277c9fc455f20d2c331ec4969389848c/spec/support/configurations_shared_context.rb#L16. Can you please see if the same error shows in tests? Is this something that changed in MongoDB 4.0 (maybe the expiration was ignored before as [the doc](https://docs.mongodb.com/manual/core/index-ttl/) says?).

**Comment by cesarizu** [comment]
I just tested with a mongodb 3.6 and it works, on 4.2 it shows that error. Running our app without the TTL index seems to work just fine, is it necessary for some functionality?

**Comment by dblock** [comment]
Yes, locks won't expire.

I think we need 

1. Add MongoDB 4.0 to CI, which should break.
2. If (1) doesn't break, add tests to ensure locks correctly expire.
3. Fix the index, I think we can have a `locking_name` sparse index and an expiring one on `_id`.

**Comment by dblock** [comment]
@cesarizu want to take a stab at any of this?

**Comment by cesarizu** [comment]
For **(1)** It seems that Travis only supports `mongodb == 3.6`. It seems that you would have to switch to Github Actions or something similar that supports `mongodb >= 4.0`.

Also about **(3)** the `_id` field doesn't support TTL indexes. I'm still wondering if the TTL index is active at all even on mongodb 3.6. A TTL index **needs** to be created on a _date_ field (according to the docs), so it would probably have to be set on the `locked_at` field.

Does that sound right?

**Comment by cesarizu** [comment]
I created a PR that fixes this issue at https://github.com/mongoid/mongoid-locker/pull/92

**Comment by cesarizu** [comment]
Thinking a bit more about this, I don't think the TTL index is a good idea. It will remove the whole document once it expires, not only the lock.

**Comment by cesarizu** [comment]
I created a branch that builds on github actions with mongo 4.2 and some other combinations. You can see the results here: https://github.com/GovSciences/mongoid-locker/actions/runs/206201310

The build is far from working completely but it shows the errors.

**Comment by dblock** [comment]
Feel free to swap travis for Github Actions! Maybe a separate PR for a first passing build?

**Comment by rsosag** [comment]
> I created a PR that fixes this issue at #92

@cesarizu so I close this issue.
